File: packages/npm/tests/client.test.mjs
import assert from "node:assert/strict";

import {
  AegirSecurityClient,
  AegirSecurityError,
  BudgetExhaustedError,
  InMemoryAgentToAgentProtocol,
  InMemoryBudgetEnvelopeService,
  signIdentityTokenPayload,
  verifyIdentityTokenPayload,
  createSafeEnvelope,
  validateSafeEnvelope
} from "../dist/esm/index.js";

async function test(name, run) {
  try {
    await run();
    console.log(`ok - ${name}`);
  } catch (error) {
    console.error(`not ok - ${name}`);
    throw error;
  }
}

await test("scanIdentity sends config and developer key headers", async () => {
  let capturedUrl = null;
  let capturedHeaders = null;

  const client = new AegirSecurityClient({
    baseUrl: "https://example.test/apis",
    configKey: "cfg_123",
    developerKey: "dev_456",
    fetcher: async (url, init) => {
      capturedUrl = url;
      capturedHeaders = new Headers(init?.headers);
      return new Response(JSON.stringify({ ok: true }), {
        status: 200,
        headers: { "Content-Type": "application/json" }
      });
    }
  });

  await client.scanIdentity("developer:dev_123");

  assert.equal(capturedUrl, "https://example.test/apis/v1/security/scan/developer%3Adev_123");
  assert.equal(capturedHeaders.get("Accept"), "application/json");
  assert.equal(capturedHeaders.get("X-Aegir-Config-Key"), "cfg_123");
  assert.equal(capturedHeaders.get("X-Aegir-Developer-Key"), "dev_456");
});

await test("performAction posts JSON payload", async () => {
  let capturedMethod = null;
  let capturedBody = null;
  let capturedContentType = null;

  const client = new AegirSecurityClient({
    baseUrl: "https://example.test/apis",
    fetcher: async (_url, init) => {
      capturedMethod = init?.method ?? "GET";
      capturedBody = init?.body;
      capturedContentType = new Headers(init?.headers).get("Content-Type");
      return new Response(JSON.stringify({ actionId: "action_123" }), {
        status: 200,
        headers: { "Content-Type": "application/json" }
      });
    }
  });

  await client.performAction({
    subjectId: "developer:dev_123",
    actionType: "LockIdentity",
    reason: "Manual review"
  });

  assert.equal(capturedMethod, "POST");
  assert.equal(capturedContentType, "application/json");
  assert.equal(capturedBody, JSON.stringify({
    subjectId: "developer:dev_123",
    actionType: "LockIdentity",
    reason: "Manual review"
  }));
});

await test("reverseAction appends encoded reason", async () => {
  let capturedUrl = null;

  const client = new AegirSecurityClient({
    baseUrl: "https://example.test/apis",
    fetcher: async (url) => {
      capturedUrl = url;
      return new Response(JSON.stringify({ actionId: "action_123" }), {
        status: 200,
        headers: { "Content-Type": "application/json" }
      });
    }
  });

  await client.reverseAction("action_123", "resolved by admin");

  assert.equal(
    capturedUrl,
    "https://example.test/apis/v1/security/actions/action_123/reverse?reason=resolved%20by%20admin"
  );
});

await test("request throws AegirSecurityError for non-200 JSON response", async () => {
  const client = new AegirSecurityClient({
    baseUrl: "https://example.test/apis",
    fetcher: async () => new Response(JSON.stringify({ error: "forbidden" }), {
      status: 403,
      headers: { "Content-Type": "application/json", "X-Correlation-Id": "corr_123" }
    })
  });

  await assert.rejects(
    () => client.getExposureSummary("developer:dev_123"),
    (error) => {
      assert.ok(error instanceof AegirSecurityError);
      assert.equal(error.status, 403);
      assert.equal(error.code, "forbidden");
      assert.equal(error.correlationId, "corr_123");
      assert.deepEqual(error.body, { error: "forbidden" });
      return true;
    }
  );
});

await test("diagnostics hooks receive log metric and trace events", async () => {
  const logs = [];
  const metrics = [];
  const traces = [];

  const client = new AegirSecurityClient({
    baseUrl: "https://example.test/apis",
    diagnosticsHooks: {
      log: (entry) => logs.push(entry),
      metric: (event) => metrics.push(event),
      trace: (event) => traces.push(event)
    },
    fetcher: async () => new Response(JSON.stringify({ graphId: "graph:developer:dev_123", generatedAtUtc: "2026-04-07T15:00:00Z", scores: { exposureScore: 0.15, trustScore: 0.92, connectionClarityScore: 0.88 }, nodes: [], edges: [], metadata: {} }), {
      status: 200,
      headers: { "Content-Type": "application/json", "X-Correlation-Id": "corr_diag" }
    })
  });

  await client.getIdentityGraph("developer:dev_123");

  assert.equal(logs[0].eventName, "request.start");
  assert.equal(metrics[0].metricName, "aegir.client.request.duration");
  assert.equal(traces.at(-1).traceName, "request.complete");
  assert.equal(traces.at(-1).correlationId, "corr_diag");
});

await test("request returns raw text body when response is not JSON", async () => {
  const client = new AegirSecurityClient({
    baseUrl: "https://example.test/apis",
    fetcher: async () => new Response("plain-text-response", {
      status: 200,
      headers: { "Content-Type": "text/plain" }
    })
  });

  const result = await client.getIdentityGraphJson("developer:dev_123");

  assert.equal(result, "\"plain-text-response\"");
});

await test("safe envelope detects tampering", async () => {
  const envelope = createSafeEnvelope("agent:test", "{\"hello\":\"world\"}");
  envelope.canonicalPayload = "{\"hello\":\"tampered\"}";
  const result = validateSafeEnvelope(envelope);
  assert.equal(result.isValid, false);
  assert.equal(result.reason, "payload_hash_mismatch");
});

await test("a2a receive rejects tampered message", async () => {
  const protocol = new InMemoryAgentToAgentProtocol();
  const message = protocol.send("agent:sender", "agent:recipient", "handoff", "{\"hello\":\"world\"}");
  message.envelope.canonicalPayload = "{\"hello\":\"tampered\"}";
  const result = protocol.receive("agent:recipient", message.messageId);
  assert.equal(result.accepted, false);
  assert.equal(result.reason, "payload_hash_mismatch");
});

await test("budget service throws exhaustion error", async () => {
  const service = new InMemoryBudgetEnvelopeService({ identityScanLimit: 1 });
  service.consume("IdentityScan", "subject-1");
  assert.throws(() => service.consume("IdentityScan", "subject-1"), (error) => {
    assert.ok(error instanceof BudgetExhaustedError);
    assert.equal(error.envelope.isExhausted, true);
    return true;
  });
});

await test("identity-token signature verifies and fails on tampering", async () => {
  const payload = new TextEncoder().encode("{\"userId\":\"user-1\",\"trustLevel\":4}");
  const signature = signIdentityTokenPayload(payload);
  assert.equal(signature.length, 64);
  assert.equal(verifyIdentityTokenPayload(payload, signature), true);

  const tampered = new TextEncoder().encode("{\"userId\":\"user-2\",\"trustLevel\":4}");
  assert.equal(verifyIdentityTokenPayload(tampered, signature), false);
});

File: packages/go/client_test.go
package aegirsecurity

import (
	"io"
	"net/http"
	"strings"
	"testing"
)

type roundTripFunc func(*http.Request) (*http.Response, error)

func (f roundTripFunc) Do(req *http.Request) (*http.Response, error) {
	return f(req)
}

func TestScanIdentity(t *testing.T) {
	client := NewAegirSecurityClient("https://example.test/apis", "", "", roundTripFunc(func(req *http.Request) (*http.Response, error) {
		return jsonResponse(200, `{"subjectId":"developer:dev_123","graph":{"graphId":"graph:developer:dev_123","generatedAtUtc":"2026-04-07T15:00:00Z","scores":{"exposureScore":0.15,"trustScore":0.92,"connectionClarityScore":0.88},"nodes":[],"edges":[],"metadata":{}},"exposure":{"graphId":"graph:developer:dev_123","generatedAtUtc":"2026-04-07T15:00:00Z","exposureScore":0.15,"findings":[],"summaryText":"Low exposure posture."},"actionHistory":[]}`), nil
	}))

	result, err := client.ScanIdentity("developer:dev_123")
	if err != nil {
		t.Fatalf("expected nil error, got %v", err)
	}

	if result.SubjectID != "developer:dev_123" {
		t.Fatalf("unexpected subject id %s", result.SubjectID)
	}
}

func TestHeadersAreAttached(t *testing.T) {
	client := NewAegirSecurityClient("https://example.test/apis", "config-key", "developer-key", roundTripFunc(func(req *http.Request) (*http.Response, error) {
		if req.Header.Get("X-Aegir-Config-Key") != "config-key" {
			t.Fatalf("missing config header")
		}
		if req.Header.Get("X-Aegir-Developer-Key") != "developer-key" {
			t.Fatalf("missing developer header")
		}
		return jsonResponse(200, `{"graphId":"graph:developer:dev_123","generatedAtUtc":"2026-04-07T15:00:00Z","scores":{"exposureScore":0.15,"trustScore":0.92,"connectionClarityScore":0.88},"nodes":[],"edges":[],"metadata":{}}`), nil
	}))

	_, err := client.GetIdentityGraph("developer:dev_123")
	if err != nil {
		t.Fatalf("expected nil error, got %v", err)
	}
}

func TestNonTwoHundredReturnsAegirSecurityError(t *testing.T) {
	client := NewAegirSecurityClient("https://example.test/apis", "", "", roundTripFunc(func(req *http.Request) (*http.Response, error) {
		response := jsonResponse(404, `{"error":"not_found"}`)
		response.Header.Set("X-Correlation-Id", "corr_123")
		return response, nil
	}))

	_, err := client.GetExposureSummary("developer:dev_123")
	if err == nil {
		t.Fatalf("expected error")
	}

	typedError, ok := err.(*AegirSecurityError)
	if !ok {
		t.Fatalf("expected AegirSecurityError, got %T", err)
	}
	if typedError.Code != "not_found" {
		t.Fatalf("expected not_found code, got %s", typedError.Code)
	}
	if typedError.CorrelationID != "corr_123" {
		t.Fatalf("expected corr_123 correlation id, got %s", typedError.CorrelationID)
	}
}

func TestDiagnosticsHooksReceiveEvents(t *testing.T) {
	logs := make([]LogEntry, 0)
	metrics := make([]MetricEvent, 0)
	traces := make([]TraceEvent, 0)

	client := NewAegirSecurityClientWithOptions(ClientOptions{
		BaseURL: "https://example.test/apis",
		HTTPClient: roundTripFunc(func(req *http.Request) (*http.Response, error) {
			response := jsonResponse(200, `{"graphId":"graph:developer:dev_123","generatedAtUtc":"2026-04-07T15:00:00Z","scores":{"exposureScore":0.15,"trustScore":0.92,"connectionClarityScore":0.88},"nodes":[],"edges":[],"metadata":{}}`)
			response.Header.Set("X-Correlation-Id", "corr_diag")
			return response, nil
		}),
		Diagnostics: &DiagnosticsHooks{
			Log: func(entry LogEntry) { logs = append(logs, entry) },
			Metric: func(event MetricEvent) { metrics = append(metrics, event) },
			Trace: func(event TraceEvent) { traces = append(traces, event) },
		},
	})

	if _, err := client.GetIdentityGraph("developer:dev_123"); err != nil {
		t.Fatalf("expected nil error, got %v", err)
	}

	if logs[0].EventName != "request.start" {
		t.Fatalf("expected request.start log, got %s", logs[0].EventName)
	}
	if metrics[0].MetricName != "aegir.client.request.duration" {
		t.Fatalf("unexpected metric %s", metrics[0].MetricName)
	}
	if traces[len(traces)-1].CorrelationID != "corr_diag" {
		t.Fatalf("expected corr_diag trace correlation id, got %s", traces[len(traces)-1].CorrelationID)
	}
}

func jsonResponse(status int, body string) *http.Response {
	return &http.Response{
		StatusCode: status,
		Body:       io.NopCloser(strings.NewReader(body)),
		Header:     make(http.Header),
	}
}
